File: services/format_answer.py
"""助手正文展示清洗：去掉 Markdown 噪音与空壳「一、二、」标题。"""
from __future__ import annotations

import re

_OUTLINE_HEADER = re.compile(r"^[一二三四五六七八九十百千]+、.+")
_CN_NUM = list("一二三四五六七八九十")
_HR = re.compile(r"^[ \t]*(?:-{3,}|\*{3,}|_{3,})[ \t]*$", re.M)
_BOLD = re.compile(r"\*\*([^*]*)\*\*")
_ITALIC = re.compile(r"\*([^*\n]+)\*")
_HEADING = re.compile(r"^#{1,6}\s+", re.M)
_MULTI_NL = re.compile(r"\n{3,}")
_NOISE_ONLY = re.compile(r"^[-–—*•·.。、：:;；\s]+$")


def _is_outline(line: str) -> bool:
    return bool(_OUTLINE_HEADER.match(line.strip()))


def _is_blank_or_noise(line: str) -> bool:
    t = line.strip()
    return (not t) or bool(_NOISE_ONLY.match(t))


def strip_empty_outline_headers(text: str) -> str:
    lines = (text or "").splitlines()
    header_idx = [i for i, ln in enumerate(lines) if _is_outline(ln)]
    drop: set[int] = set()
    for k, start in enumerate(header_idx):
        end = header_idx[k + 1] if k + 1 < len(header_idx) else len(lines)
        body = lines[start + 1 : end]
        has_content = any(
            (not _is_blank_or_noise(ln)) and (not _is_outline(ln)) for ln in body
        )
        if not has_content:
            drop.add(start)

    kept = [ln for i, ln in enumerate(lines) if i not in drop]
    n = 0
    out: list[str] = []
    for ln in kept:
        if _is_outline(ln) and n < len(_CN_NUM):
            ln = re.sub(
                r"^([ \t]*)[一二三四五六七八九十百千]+、",
                rf"\1{_CN_NUM[n]}、",
                ln,
                count=1,
            )
            n += 1
        out.append(ln)
    return _MULTI_NL.sub("\n\n", "\n".join(out)).strip()


def clean_assistant_text(text: str) -> str:
    """落库/返回前清洗助手正文。"""
    if not text:
        return ""
    s = str(text)
    s = _BOLD.sub(r"\1", s)
    s = _ITALIC.sub(r"\1", s)
    s = _HEADING.sub("", s)
    s = s.replace("**", "")
    s = _HR.sub("", s)
    s = _MULTI_NL.sub("\n\n", s)
    s = strip_empty_outline_headers(s)
    return s.strip()

File: frontend/src/utils/formatAnswer.js
/**
 * 展示前去掉 LLM 常见的 Markdown 噪音（不做完整 MD 渲染）。
 * 目标：气泡里是可读正文，而不是 `---` / 空壳标题等原稿痕迹。
 */
export function stripMarkdownInline(text) {
  if (!text) return ''
  let s = String(text)
    // 行内加粗/斜体
    .replace(/\*\*([^*]*)\*\*/g, '$1')
    .replace(/\*([^*\n]+)\*/g, '$1')
    .replace(/^#{1,6}\s+/gm, '')
    .replace(/\*\*/g, '')
    // Markdown 分隔线（单独一行的 --- / *** / ___）
    .replace(/^[ \t]*(?:-{3,}|\*{3,}|_{3,})[ \t]*$/gm, '')
    // 多余空行收成最多一个空行
    .replace(/\n{3,}/g, '\n\n')

  s = stripEmptyOutlineHeaders(s)
  return s.trim()
}

/** 中文大标题：一、二、三…（不含「1. 正文」这种同条列表） */
const OUTLINE_HEADER = /^[一二三四五六七八九十百千]+、.+/
const CN_NUM = ['一', '二', '三', '四', '五', '六', '七', '八', '九', '十']

function isOutlineHeader(line) {
  return OUTLINE_HEADER.test(String(line || '').trim())
}

/** 空行，或只有标点/项目符号、没有实质正文 */
function isBlankOrNoise(line) {
  const t = String(line || '').trim()
  if (!t) return true
  if (/^[-–—*•·.。、：:;；\s]+$/.test(t)) return true
  return false
}

/**
 * 去掉「标题下没有实质正文」的空壳大纲标题，并对剩余标题重新编号。
 * 例如只剩「三、注意事项」→「一、注意事项」。
 */
export function stripEmptyOutlineHeaders(text) {
  const lines = String(text || '').split(/\r?\n/)
  const headerIdx = []
  for (let i = 0; i < lines.length; i++) {
    if (isOutlineHeader(lines[i])) headerIdx.push(i)
  }

  const drop = new Set()
  for (let k = 0; k < headerIdx.length; k++) {
    const start = headerIdx[k]
    const end = k + 1 < headerIdx.length ? headerIdx[k + 1] : lines.length
    const body = lines.slice(start + 1, end)
    const hasContent = body.some((l) => !isBlankOrNoise(l) && !isOutlineHeader(l))
    if (!hasContent) drop.add(start)
  }

  const kept = lines.filter((_, i) => !drop.has(i))
  let n = 0
  const renumbered = kept.map((line) => {
    if (!isOutlineHeader(line) || n >= CN_NUM.length) return line
    return line.replace(
      /^([ \t]*)[一二三四五六七八九十百千]+、/,
      `$1${CN_NUM[n++]}、`
    )
  })
  return renumbered.join('\n').replace(/\n{3,}/g, '\n\n').trim()
}
